core/src/main/java/smile/classification/DecisionTree.java

```diff
@@ -982,6 +997,119 @@ public DecisionTree(Attribute[] attributes, double[][] x, int[] y, int maxNodes,
+                        if (samples[variableOrder[j]] != 0) {
+                            compressedVariableOrder[k++] = variableOrder[j];
+                        }
+                    }
+                    compressedOrder[i] = compressedVariableOrder;
+                }
+            }
+            order = compressedOrder;
+
+            originalOrder = new int[presentCount];
+            int k = 0;
+            for (int i = 0; i < n; i++) {
+                if (samples[i] != 0) {
+                    originalOrder[k++] = i;
+                }
+            }
+        } else {
+            originalOrder = new int[n];
+            for (int i = 0; i < n; i++) {
+                originalOrder[i] = i;
+            }
+        }
+    }
+
+    /**
+     * Modifies {@link #order} and {@link #originalOrder} by partitioning the range from low
+     * (inclusive) to high (exclusive) so that all elements o for which goesLeft(o) is true come
+     * before all elements for which it is false, but element ordering is otherwise preserved. The
+     * number of true elements in left must equal split-low.
+     * @param low the low bound of the segment of the order arrays which will be partitioned.
+     * @param split where the partition's split point will end up.
+     * @param high the high bound of the segment of the order arrays which will be partitioned.
+     * @param goesLeft whether an element goes to the left side or the right side of the
+     *        partition.
+     * @param buffer scratch space large enough to hold all elements for which left is false.
+     */
+    private void partitionOrder(int low, int split, int high, IntPredicate goesLeft, int[] buffer) {
+        for (int[] variableOrder : order) {
+            if (variableOrder != null) {
+                partitionArray(variableOrder, low, split, high, goesLeft, buffer);
+            }
+        }
+        partitionArray(originalOrder, low, split, high, goesLeft, buffer);
+    }
+
+    /**
+     * Modifies an array in-place by partitioning the range from low (inclusive) to high
+     * (exclusive) so that all elements i for which left[i - low] is true come before all elements
+     * for which it is false, but element ordering is otherwise preserved. The number of true
+     * elements in left must equal split-low. buffer is scratch space large enough to hold all
+     * elements for which left is false.
+     */
+    private void partitionArray(int[] a, int low, int split, int high, IntPredicate goesLeft, int[] buffer) {
+        int n = high - low;
+        int j = low;
+        int k = 0;
+        for (int i = low; i < high; i++) {
+            if (goesLeft.test(a[i])) {
+                a[j++] = a[i];
+            } else {
+                buffer[k++] = a[i];
+            }
+        }
+        if (k != high - split || j != split) {
+            System.err.println("Messed up partition: " + low + ".." + split + ".." + high + " ended up splitting at " + j);
```

Comment:
Better to throw IllegalStateException for fail-fast if this condition should not be met.

Reply:
Sigh. There's always one left-over bit of debugging that should have been removed. The unit tests do pass.